Fix editor test play not marking hit objects before its start time as judged

LeNitrous opened the pull request:
Fixes https://github.com/ppy/osu/issues/17395 and https://github.com/ppy/osu/issues/23029 using the suggestion from https://github.com/ppy/osu/issues/17395#issuecomment-1075749829.


https://github.com/ppy/osu/assets/20495991/fc8c02f1-6c3e-400d-8d11-ed0f28769090

LeNitrous committed:
Make all hit objects before the start time marked as hit.

LeNitrous committed:
Enumerate nested hit objects

LeNitrous committed:
add test case

turbedi commented on osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs:
```diff
@@ -43,6 +51,22 @@ protected override GameplayClockContainer CreateGameplayClockContainer(WorkingBe
+            {
+                var judgement = hitObject.CreateJudgement();
+
+                if (!frame.Header.Statistics.ContainsKey(judgement.MaxResult))
+                    frame.Header.Statistics.Add(judgement.MaxResult, 0);
```
```suggestion
                frame.Header.Statistics.TryAdd(judgement.MaxResult, 0);
```

LeNitrous committed:
`Contains` + `Add` to `TryAdd`

frenzibyte commented on osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs:
```diff
@@ -43,6 +51,19 @@ protected override GameplayClockContainer CreateGameplayClockContainer(WorkingBe
+
+            var frame = new ReplayFrame { Header = new FrameHeader(new ScoreInfo(), new ScoreProcessorStatistics()) };
+
+            foreach (var hitObject in enumerateHitObjects(playableBeatmap.HitObjects.Where(h => h.StartTime < editorState.Time)))
```
Just use `DrawableRuleset.Objects`?

frenzibyte reviewed (dismissed):
osu!catch has automagically been working prior to this PR due to the nature of its hit object judgement logic (judging all object prior to current time as miss, instead of ignoring).

With this PR, now the judged hits count goes over the maximum value, as all previous objects increase the judged hits counter with "miss" judgements after the score processor has already been set to the calculated judged hits in this PR. This results in the osu!catch editor test player not being able to exit at finish (whereas on master it works).

An alternative method I have in mind would be to give maximum results to the actual hit objects of the test session themselves (which are assigned as entries to the drawable hit objects), so that they're not judged with a "miss" later on. I tried doing that, but I cannot seem to get it work at first attempt...something along the lines of the following diff:
<details>
<summary>diff</summary>

```diff
diff --git a/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs b/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs
index e2b2b067e0..f8fdb1f809 100644
--- a/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs
+++ b/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
 using osu.Framework.Allocation;
@@ -8,6 +9,7 @@
 using osu.Game.Beatmaps;
 using osu.Game.Online.Spectator;
 using osu.Game.Overlays;
+using osu.Game.Rulesets.Judgements;
 using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Replays;
 using osu.Game.Rulesets.Scoring;
@@ -21,19 +23,17 @@ public partial class EditorPlayer : Player
     {
         private readonly Editor editor;
         private readonly EditorState editorState;
-        private readonly IBeatmap playableBeatmap;
 
         protected override UserActivity InitialActivity => new UserActivity.TestingBeatmap(Beatmap.Value.BeatmapInfo, Ruleset.Value);
 
         [Resolved]
         private MusicController musicController { get; set; } = null!;
 
-        public EditorPlayer(Editor editor, IBeatmap playableBeatmap)
+        public EditorPlayer(Editor editor)
             : base(new PlayerConfiguration { ShowResults = false })
         {
             this.editor = editor;
             editorState = editor.GetState();
-            this.playableBeatmap = playableBeatmap;
         }
 
         protected override GameplayClockContainer CreateGameplayClockContainer(WorkingBeatmap beatmap, double gameplayStart)
@@ -52,21 +52,16 @@ protected override void LoadComplete()
         {
             base.LoadComplete();
 
-            var frame = new ReplayFrame { Header = new FrameHeader(new ScoreInfo(), new ScoreProcessorStatistics()) };
-
-            foreach (var hitObject in enumerateHitObjects(playableBeatmap.HitObjects.Where(h => h.StartTime < editorState.Time)))
+            foreach (var hitObject in enumerateHitObjects(DrawableRuleset.Objects))
             {
-                var judgement = hitObject.CreateJudgement();
-
-                if (!frame.Header.Statistics.ContainsKey(judgement.MaxResult))
-                    frame.Header.Statistics.Add(judgement.MaxResult, 0);
+                if (hitObject.StartTime >= editorState.Time)
+                    continue;
 
-                frame.Header.Statistics[judgement.MaxResult]++;
+                var judgement = hitObject.CreateJudgement();
+                HealthProcessor.ApplyResult(new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult });
+                ScoreProcessor.ApplyResult(new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult });
             }
 
-            HealthProcessor.ResetFromReplayFrame(frame);
-            ScoreProcessor.ResetFromReplayFrame(frame);
-
             ScoreProcessor.HasCompleted.BindValueChanged(completed =>
             {
                 if (completed.NewValue)
```

</summary>

frenzibyte commented:
In addition, there also seem to be specific scenarios at which editor test play still fails to exit on finish:

https://github.com/ppy/osu/assets/22781491/d4701306-e106-403a-bb17-72c98e507872

At this point it may seem more ideal to put a logic somewhere to exit when passing by the last hit object or the end of the track. Something to keep in mind as a better alternative.

bdach commented:
> At this point it may seem more ideal to add local code somewhere to mark gameplay as "completed" when being 1000ms ahead of the last hit object end time. Something to keep in mind as a better alternative.

disagree. i would like an explanation as to why this method is failing before resorting to such hacks

frenzibyte commented:
Hence why I mentioned it as an alternative.

bdach commented:
i don't consider it a valid alternative.

frenzibyte commented:
I will wait for a response to the issues I pointed above to get the current method in a fully working state, and then discuss it from there.

LeNitrous commented:
> > At this point it may seem more ideal to add local code somewhere to mark gameplay as "completed" when being 1000ms ahead of the last hit object end time. Something to keep in mind as a better alternative.
> 
> disagree. i would like an explanation as to why this method is failing before resorting to such hacks

It was caused by not accounting for hit windows and hit objects with a duration causing hit objects to be judged twice making `JudgedHits > MaxHits` in `JudgementProcessor`.

> osu!catch has automagically been working prior to this PR due to the nature of its hit object judgement logic (judging all object prior to current time as miss, instead of ignoring).

Other modes were working as well. It's because in other modes other than standard, they mark all hit objects after a certain time as missed causing `JudgedHits * 2 ≈ MaxHits` in this PR.

peppy committed:
Merge branch 'master' into fix/editor-test-exit

peppy committed:
Avoid passing beatmap in from editor when it's already present

peppy committed:
Fix the majority of cases where gameplay stil doesn't end due to judgement count mismatch

peppy committed:
Apply results rather than fake a replay frame

peppy committed:
Fix incorrect handling of nested objects when inside parent object's duration

peppy commented:
I've fixed multiple edge cases and applied @frenzibyte's proposed changes above (applying results instead of using replay frame) as it solves the issue of DHOs currently on screen duplicating judgements (this could not be resolved with the replay frame approach).

Not sure if we want to try and craft tests for this but before expending the effort in doing so I'd like to see if anyone can break this.

bdach commented:
> this could not be resolved with the replay frame approach

Can you elaborate on this any further?

I guess I'll have a look at the end of the day but I'm _really_ loath to be inventing another way to do this. Especially so that I don't see how the replay frame mechanism can work for multi but not for editor. I will need to find out myself why that is the case before passing any judgement (whether there are some special editor circumstances at play, or whether the multi spectator mechanism is broken e.g. for catch and needs to be either fixed or replaced).

peppy commented:
> Can you elaborate on this any further?

This is correctly touched on by frenzi above, but I'll attempt.

Basically, assume there are multiple DHOs on screen, some before `test_start_time` and some after. When only using `HitObjects` to create the replay frame judgement count, we have no easy way of telling which objects are on screen (and will trigger their own `ApplyResult(miss)`). We would need to use some concoction of `PreEmpt` or transforms or something else to get a correct count for the current frame of entry (and even then I'm not sure we could be 100% accurate).

In a simple scenario where past objects are not mapped to DHOs – ie. only objects after `test_start_time` exist – there will be no issue.

Which is to say, I honestly don't know how this works in replays (it probably doesn't always).

bdach commented on osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs:
```diff
@@ -56,6 +62,32 @@ protected override void LoadComplete()
 
+        private void markPreviousObjectsHit()
+        {
+            foreach (var hitObject in enumerateHitObjects(DrawableRuleset.Objects, editorState.Time))
+            {
+                var judgement = hitObject.Judgement;
+
+                HealthProcessor.ApplyResult(new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult });
+                ScoreProcessor.ApplyResult(new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult });
+            }
```
If the problem with the replay frame-based approach is flawed because it relied on using hitobjects and hitobject-based cutoffs, then how is _this_ fine? Why is there no attempt to compensate for preempt or transforms or whatever?

That explanation isn't making much sense to me in light of this.

peppy replied:
The hookup flow with `ScoreProcessor` via `Player` means that applying result here will have the results end up on the DHOs, seemingly. It's a long chain of events that bugs me every time i have to traverse it.

bdach commented:
> I'd like to see if anyone can break this.

Started https://osu.ppy.sh/beatmapsets/310607#mania/710980 on 01:37:623 and it kicked me out of test play before I actually completed the map:

https://github.com/ppy/osu/assets/20418176/d5304710-f82e-4835-ac7e-76aa8c1a22ec

peppy commented:
👍 safe to say this is all doomed then, and go back to the drawing board.

bdach commented:
Coming back to this: As far as I can tell, the assertions [from this comment](https://github.com/ppy/osu/pull/26465#discussion_r1620209171) that going through `ScoreProcessor` does something with DHOs is pretty clearly false.

However, how do you feel about the following?

```diff
diff --git a/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs b/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs
index 2028094964..9a7c1822a3 100644
--- a/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs
+++ b/osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs
@@ -48,6 +48,7 @@ protected override void LoadComplete()
             base.LoadComplete();
 
             markPreviousObjectsHit();
+            markVisibleDrawableObjectsHit();
 
             ScoreProcessor.HasCompleted.BindValueChanged(completed =>
             {
@@ -67,9 +68,10 @@ private void markPreviousObjectsHit()
             foreach (var hitObject in enumerateHitObjects(DrawableRuleset.Objects, editorState.Time))
             {
                 var judgement = hitObject.Judgement;
+                var result = new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult };
 
-                HealthProcessor.ApplyResult(new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult });
-                ScoreProcessor.ApplyResult(new JudgementResult(hitObject, judgement) { Type = judgement.MaxResult });
+                HealthProcessor.ApplyResult(result);
+                ScoreProcessor.ApplyResult(result);
             }
 
             static IEnumerable<HitObject> enumerateHitObjects(IEnumerable<HitObject> hitObjects, double cutoffTime)
@@ -88,6 +90,40 @@ static IEnumerable<HitObject> enumerateHitObjects(IEnumerable<HitObject> hitObje
             }
         }
 
+        private void markVisibleDrawableObjectsHit()
+        {
+            if (!DrawableRuleset.Playfield.IsLoaded)
+            {
+                Schedule(markVisibleDrawableObjectsHit);
+                return;
+            }
+
+            foreach (var drawableObjectEntry in enumerateDrawableEntries(
+                         DrawableRuleset.Playfield.AllHitObjects
+                                        .Select(ho => ho.Entry)
+                                        .Where(e => e != null)
+                                        .Cast<HitObjectLifetimeEntry>(), editorState.Time))
+            {
+                drawableObjectEntry.Result = new JudgementResult(drawableObjectEntry.HitObject, drawableObjectEntry.HitObject.Judgement)
+                    { Type = drawableObjectEntry.HitObject.Judgement.MaxResult };
+            }
+
+            static IEnumerable<HitObjectLifetimeEntry> enumerateDrawableEntries(IEnumerable<HitObjectLifetimeEntry> entries, double cutoffTime)
+            {
+                foreach (var entry in entries)
+                {
+                    foreach (var nested in enumerateDrawableEntries(entry.NestedEntries, cutoffTime))
+                    {
+                        if (nested.HitObject.GetEndTime() < cutoffTime)
+                            yield return nested;
+                    }
+
+                    if (entry.HitObject.GetEndTime() < cutoffTime)
+                        yield return entry;
+                }
+            }
+        }
+
         protected override void PrepareReplay()
         {
             // don't record replays.

```

Is it pretty cursed? Yeah. Does it appear to work? In my (somewhat brief) testing, *also* yeah. It should also scale to non-pooled rulesets, thankfully, because of that whole `SyntheticHitObjectEntry` shebang that is present on non-pooled objects.

As far as I can tell the issue happens because the DHOs do not get the `Result` on their lifetime entry adjusted in line with what was faked on the score/HP processors. Therefore, especially in cases like osu! sliders or mania holds, it can occur that the DHO is not aware that it has already been accounted for in the judgement processors, and as such will issue judgement results itself, which leads to either ending gameplay early (if it causes the processors to hit `JudgedHits == MaxHits` prematurely), or not ending at all (if it causes the processors' `JudgedHits` to _exceed_ `MaxHits`).

peppy commented:
I think the diff looks okay.

bdach committed:
Merge branch 'master' into fix/editor-test-exit

bdach committed:
Expand test coverage

- Covers fail case that wasn't covered before
- Removes arbitrary wait step that was inevitably going to cause
  intermittent test failures

bdach committed:
Ensure past drawable objects also get their results populated in editor test play

bdach committed:
Fix formatting
